### ISSUE TITLE
Proposal: split non-strict relevance into shape-irrelevance, parametricity, and runtime-irrelevance

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#2757]
Right now, we're using the non-strict `..` modality for (at least) three purposes:

- The codomain of an irrelevant function type must depend non-strictly on its argument (aka shape-irrelevance).
- Variables that are bound non-strictly can be used relevantly at the type level but only in a 'parametric' way at the term level (aka parametricity).
- Non-strict arguments can be erased during compilation (aka runtime-irrelevance).

I think Agda should make the distinction between these three different concepts. 

A possible idea is to profit from the work already done by @Saizan on the parametric branch and cherry-pick some of the changes there to the stable branch (in particular the `#` modality, but not the new primitives). This would also reduce the burden of maintaining that branch, and make it easier to merge in the future (if we decide to do that).

**Comment by nad** [comment]
> I think Agda should make the distinction between these three different concepts.

I suggest that we use the following notations: `....`, `.....` and `......`. (I don't want to overload `...`, and code using `..` should trigger an error.) I guess I don't need to explain what modality each of the notations refers to.

**Comment by jespercockx** [comment]
That's a great proposal, however on the parametric branch there's actually something else already implemented. For example, you can write `(x :{#} A)` to bind x parametrically.

**Comment by jespercockx** [comment]
Oh, and something else: even keeping track of parametricity internally without assigning it a specific syntax would be helpful, as parametricity can be 'created' by going from the term level to the type level (i.e. by `workOnTypes`). But if we're putting in the effort of implementing the guts, I'd very much like to annotate it explicitly, too. The idea is that this change would be mostly backwards-compatible (well, apart from the current abuses of the `..` modality), so you can keep writing your Agda code without extra annotations if you don't need shape-irrelevance/parametricity/runtime-irrelevance.

**Comment by UlfNorell** [comment]
There's some syntax discussion in #2513. I'm not really fond of the `(x :{#} A)` syntax since it doesn't handle `∀`-bindings and separates the variable from its type too much (especially if the modality annotation is longer than a couple of characters).

**Comment by andreasabel** [comment]
At least for function types and typed bindings, the following syntax is available:
```
  {p} A -> B
  (x y z : {p} A) -> B
```
for modality `p` in function type.

**Comment by andreasabel** [comment]
@nad: For the record, using `.` for irrelevance was your idea back then...

**Comment by jespercockx** [comment]
Please, let's keep the discussion about syntax in #2513 and keep the discussion here about the *internal* representation of relevances/modalities.

A question related to the discussion here is whether we actually need a separate annotation for runtime irrelevance. If irrelevance and shape-irrelevance were working properly, wouldn't it be better to make those arguments _actually_ irrelevant? Or is there some use-case of runtime irrelevance that I don't see and that cannot be replaced by genuine irrelevance?

**Comment by UlfNorell** [comment]
Compile-time and run-time irrelevance are clearly different (`Vec A n != Vec A m`, but you might not need to carry the length at run-time). Shape irrelevance is closer, which is why they're currently collapsed, but the fact that @andreasabel calls this an abuse suggests that it isn't a perfect match...

**Comment by jespercockx** [comment]
Well, if you make `Vec` shape-irrelevant in the length, then you can make its constructors actually irrelevant in the length. The result is that `Vec A n != Vec A m`, but the length is erased from the constructors both at compile time **and** at runtime. At the type-level, the index is not erased but this shouldn't be a problem since types can be erased altogether. So is there still some reason to use runtime irrelevance instead of this approach?

**Comment by UlfNorell** [comment]
I wasn't thinking about the constructors, but functions over vectors. I'd like to be able to get rid of the length argument to vector map for instance. What modality would I use for that?

**Comment by jespercockx** [comment]
You can also use real irrelevance:
```agda
{-# OPTIONS --experimental-irrelevance #-}
open import Agda.Builtin.Nat

data Vec (A : Set) : ..(n : Nat) → Set where
  []  : Vec A 0
  _∷_ : .{n : Nat} → A → Vec A n → Vec A (suc n)

map : {A B : Set} (f : A → B) .{n : Nat} → Vec A n → Vec B n
map f []       = []
map f (x ∷ xs) = f x ∷ map f xs
```
Currently, this still needs the `--experimental-irrelevance` option, but the idea is that this becomes the default.

The thing that's currently possible with the `..` modality that would *not* be possible after the split is to use a `..` argument relevantly at the type level:
```agda
-- currently possible, but would become invalid
foo : ..{A : Set} → A → A
foo x = x
```
Instead, here you will be able to choose between making A either relevant or parametric:
```agda
-- also possible, will stay valid (obviously)
foo₁ : {A : Set} → A → A
foo₁ x = x

-- new option: make foo parametric in A
foo₂ : {A :{#} Set} → A → A
foo₂ x = x
```

**Comment by nad** [comment]
> @nad: For the record, using `.` for irrelevance was your idea back then...

Really? I have no recollection of this.

**Comment by UlfNorell** [comment]
@jespercockx:

Can you elaborate what the problem is when trying to do what I was doing before the fix of #2640:

```agda
open import Agda.Builtin.Nat

data Vec (A : Set) : Nat → Set where
  []  : Vec A zero
  _∷_ : ∀ ..{n} → A → Vec A n → Vec A (suc n)

map : ∀ {A B n} → (A → B) → Vec A n → Vec B n
map f []       = []
map f (x ∷ xs) = f x ∷ map f xs
-- I'm not sure if there should be a case for the constructor _∷_,
-- because I get stuck when trying to solve the following unification
-- problems (inferred index ≟ expected index):
--   suc n₁ ≟ n
-- when checking that the pattern x ∷ xs has type Vec A n
```
Having to add a modality to the length argument of `Vec` is really inconvenient, and being able to erase the length argument of `_∷_` is crucial for performance, so it would be very nice if we could make this work.

**Comment by jespercockx** [comment]
Now that we have the `@0` modality, all that's left to do is to make the `..` modality no longer imply run-time irrelevance.
